File: api/v1/schemas/market_review.py
# -*- coding: utf-8 -*-
"""
===================================
大盘复盘接口 Schema
===================================
"""

from typing import Literal, Optional

from pydantic import BaseModel, Field

RegionType = Literal["cn", "hk", "us", "both", "all"]
TodayRegionType = Literal["cn", "hk"]


class MarketReviewRunRequest(BaseModel):
    region: RegionType = Field("cn", description="市场区域: cn=A股, hk=港股, us=美股, both=A股+美股, all=全部")


class MarketReviewTaskAccepted(BaseModel):
    task_id: str
    status: str = "pending"
    region: str
    created_at: str


class MarketReviewStatusResponse(BaseModel):
    task_id: str
    status: str  # pending / processing / completed / failed
    region: str
    report: Optional[str] = None
    error: Optional[str] = None
    created_at: str
    completed_at: Optional[str] = None


class MarketReviewTodayResponse(BaseModel):
    region: TodayRegionType
    trade_date: Optional[str] = None
    report: Optional[str] = None
    created_at: Optional[str] = None

File: apps/dsa-web/src/api/marketReview.ts
import apiClient from "./index";

export type MarketReviewRegion = "cn" | "hk" | "us" | "both" | "all";
export type MarketReviewTodayRegion = "cn" | "hk";

export type MarketReviewTaskAccepted = {
  taskId: string;
  status: string;
  region: string;
  createdAt: string;
};

export type MarketReviewStatus = {
  taskId: string;
  status: "pending" | "processing" | "completed" | "failed";
  region: string;
  report: string | null;
  error: string | null;
  createdAt: string;
  completedAt: string | null;
};

export type MarketReviewToday = {
  region: MarketReviewTodayRegion;
  tradeDate: string | null;
  report: string | null;
  createdAt: string | null;
};

function toCamel<T>(obj: Record<string, unknown>): T {
  const result: Record<string, unknown> = {};
  for (const [k, v] of Object.entries(obj)) {
    const camel = k.replace(/_([a-z])/g, (_, c: string) => c.toUpperCase());
    result[camel] = v;
  }
  return result as T;
}

export const marketReviewApi = {
  async run(
    region: MarketReviewRegion = "cn",
  ): Promise<MarketReviewTaskAccepted> {
    const response = await apiClient.post<Record<string, unknown>>(
      "/api/v1/market-review/run",
      { region },
      { validateStatus: (s) => s === 202 || s === 200 },
    );
    return toCamel<MarketReviewTaskAccepted>(response.data);
  },

  async getStatus(taskId: string): Promise<MarketReviewStatus> {
    const response = await apiClient.get<Record<string, unknown>>(
      `/api/v1/market-review/status/${taskId}`,
    );
    return toCamel<MarketReviewStatus>(response.data);
  },

  async getToday(
    region: MarketReviewTodayRegion = "cn",
  ): Promise<MarketReviewToday> {
    const response = await apiClient.get<Record<string, unknown>>(
      "/api/v1/market-review/today",
      { params: { region } },
    );
    return toCamel<MarketReviewToday>(response.data);
  },
};
